lerobot/common/robot_devices/cameras/ros2.py

```diff
@@ -0,0 +1,277 @@
+
+    images_dir = Path(images_dir)
+    if images_dir.exists():
+        shutil.rmtree(
+            images_dir,
+        )
+    images_dir.mkdir(parents=True, exist_ok=True)
+
+    print(f"Saving images to {images_dir}")
+    frame_index = 0
+    start_time = time.perf_counter()
+    with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
+        while True:
+            now = time.perf_counter()
+
+            for camera in cameras:
+                # If we use async_read when fps is None, the loop will go full speed, and we will endup
+                # saving the same images from the cameras multiple times until the RAM/disk is full.
+                image = camera.read()
+                executor.submit(
+                    save_image,
+                    image,
+                    config.topic,
+                    frame_index,
+                    images_dir,
+                )
+
+            print(f"Frame: {frame_index:04d}\tLatency (ms): {(time.perf_counter() - now) * 1000:.2f}")
+
+            if time.perf_counter() - start_time > record_time_s:
+                break
+
+            frame_index += 1
+
+    print(f"Images have been saved to {images_dir}")
+
+
+class ROS2Camera:
+    """
+    The ROS2Camera class allows to record images from cameras over ROS 2.
+
+    An ROS2Camera instance requires a ROS 2 driver for the camera to be running with images published over a known topic, eg. /camera/image.
+    It is therefore compatible with any camera that has a ROS 2 driver which may be running on the same host or any other host on the same network.
+
+    To find the topic published by the ROS 2 driver, you may list the various topics in your network:
+    ```bash
+    # Make sure to source your ROS 2 installation first.
+    ros2 topic list
+    ```
+
+    Example of usage:
+    ```python
+    from lerobot.common.robot_devices.cameras.configs import ROS2CameraConfig
+
+    config = ROS2CameraConfig(node.name=0)
+    camera = ROS2Camera(config)
+    camera.connect()
+    color_image = camera.read()
+    # when done using the camera, consider disconnecting
+    camera.disconnect()
+    ```
+
+    Example of configuring the camera:
+
+    Note: Apart from the topic name, all other options are only configurable
+    on the ROS 2 driver side. We could infer some of these settings once we
+    successfully subscribe to the camera topic but some layers of the lerobot
+    stack (eg. recording an episode) require these attributes to be present in
+    the CameraConfig implementation before the first frame is read from the
+    camera. So for now, the user needs to ensure the settings here match those
+    on the ROS 2 driver side.
+
+    ```python
+    config = ROS2CameraConfig(topic="/image_raw", fps=30, width=640, height=480)
+    # Note: might error out open `camera.connect()` if these settings are not compatible with the camera
+    ```
+    """
+
+    def __init__(self, config: ROS2CameraConfig):
+        rclpy.init()
```

Comment:
Would be good if we can support multiple cameras here. So one way is to init the node if it isnt created.

```
        if not rclpy.ok():
            rclpy.init()
            self.node = rclpy.create_node("lerobot_camera_node")
            self.thread = threading.Thread(target=rclpy.spin, args=(self.node,), daemon=True)
            self.thread.start()
        else:
            self.thread = None
            executor = rclpy.get_global_executor()
            self.node = executor.get_nodes()[0]  # will only take the first node
```